armsrc/em4x50.c

```diff
@@ -0,0 +1,877 @@
+
+    tag.sectors[pos][5] = bits2byte(col_parity,8);
+    
+    // stop bit
+    tag.sectors[pos][6] = bits[44];
+}
+
+static void wait_timer(int timer, uint32_t period) {
+
+    // do nothing for <period> using timer <timer>
+    
+    if (timer == FPGA_TIMER_0) {
+
+        AT91C_BASE_TC0->TC_CCR = AT91C_TC_SWTRG;
+        while (AT91C_BASE_TC0->TC_CV < period);
+
+    } else {
+
+        AT91C_BASE_TC1->TC_CCR = AT91C_TC_SWTRG;
+        while (AT91C_BASE_TC1->TC_CV < period);
+
+    }
+}
+
+static void em4x50_setup_read(void) {
+ 
+    FpgaDownloadAndGo(FPGA_BITSTREAM_LF);
+    
+    FpgaWriteConfWord(FPGA_MAJOR_MODE_LF_ADC | FPGA_LF_ADC_READER_FIELD);
+
+    // 50ms for the resonant antenna to settle.
+    SpinDelay(50);
+    // Now set up the SSC to get the ADC samples that are now streaming at us.
+    FpgaSetupSsc();
+    // start a 1.5ticks is 1us
+    StartTicks();
+ 
+    FpgaSendCommand(FPGA_CMD_SET_DIVISOR, LF_DIVISOR_125);
+
+    // Connect the A/D to the peak-detected low-frequency path.
+    SetAdcMuxFor(GPIO_MUXSEL_LOPKD);
+    
+    // Steal this pin from the SSP (SPI communication channel with fpga) and
+    // use it to control the modulation
+    AT91C_BASE_PIOA->PIO_PER = GPIO_SSC_DOUT;
+    AT91C_BASE_PIOA->PIO_OER = GPIO_SSC_DOUT;
+
+    // Disable modulation at default, which means enable the field
+    LOW(GPIO_SSC_DOUT);
+    
+    // Enable Peripheral Clock for
+    //   TIMER_CLOCK0, used to measure exact timing before answering
+    //   TIMER_CLOCK1, used to capture edges of the tag frames
+    AT91C_BASE_PMC->PMC_PCER |= (1 << AT91C_ID_TC0) | (1 << AT91C_ID_TC1);
+    AT91C_BASE_PIOA->PIO_BSR = GPIO_SSC_FRAME; // wofür ist das?
+
+    // Disable timer during configuration
+    AT91C_BASE_TC0->TC_CCR = AT91C_TC_CLKDIS;
+    AT91C_BASE_TC1->TC_CCR = AT91C_TC_CLKDIS;
+
+    // TC0: Capture mode, default timer source = MCK/2 (TIMER_CLOCK1), no triggers
+    AT91C_BASE_TC0->TC_CMR = AT91C_TC_CLKS_TIMER_DIV1_CLOCK;
+    
+    // TC1: Capture mode, default timer source = MCK/2 (TIMER_CLOCK1), no triggers
+    AT91C_BASE_TC1->TC_CMR = AT91C_TC_CLKS_TIMER_DIV1_CLOCK;
+      
+    // Enable and reset counters
+    AT91C_BASE_TC0->TC_CCR = AT91C_TC_CLKEN | AT91C_TC_SWTRG;
+    AT91C_BASE_TC1->TC_CCR = AT91C_TC_CLKEN | AT91C_TC_SWTRG;
+    
+    // synchronized startup procedure
+    while (AT91C_BASE_TC0->TC_CV > 0) {}; // wait until TC1 returned to zero
+    
+    // Watchdog hit
+    WDT_HIT();
+}
+
+// functions for "reader" use case
+
+static int get_next_bit(void) {
+    
+    // returns bit value (or EM4X50_BIT_OTHER -> no bit pattern) by evaluating
+    // a single sample within a bit period (given there is no LIW, ACK or NAK)
+    // This function is not used for decoding, it is only used for identifying
+    // a listen window (return value = EM4X50_BIT_OTHER) in functions
+    // "find_double_listen_window" and "check_ack"
+  
+    int dhigh = 230;    // 90% of maximum sample value (255)
```

Comment:
these kinds of envelope values would need to be dynamic in order to get better distances.

Reply:
What do you mean by "distances"? Would a #define statement regarding "dhigh" at the top of the file be sufficient?

Reply:
No,   with distances I mean read/write distance between tag and antenna.
Having a hardcoded value means you limit it. 

Usually if you sample the signal strength  32 or 64 times, you can get an average of signal strength.
that would be your "HI" value,  from that you calc your envelope trigger 90% value,    hence allowing for much better distances.

Reply:
That does make sense.
I added a EM4x50 specific function for this calculation, but I'm still using two hardcoded values ("noise", "sample_ref"). From my point of view these two remaining values don't have to be set dynamically. ;-)